File: leetcode/bit-manipulation/645.错误的集合.ts
/*
 * @lc app=leetcode.cn id=645 lang=typescript
 *
 * [645] 错误的集合
 *
 * https://leetcode-cn.com/problems/set-mismatch/description/
 *
 * algorithms
 * Easy (33.69%)
 * Likes:    86
 * Dislikes: 0
 * Total Accepted:    12.8K
 * Total Submissions: 30.6K
 * Testcase Example:  '[1,2,2,4]'
 *
 * 集合 S 包含从1到 n
 * 的整数。不幸的是，因为数据错误，导致集合里面某一个元素复制了成了集合里面的另外一个元素的值，导致集合丢失了一个整数并且有一个元素重复。
 *
 * 给定一个数组 nums 代表了集合 S 发生错误后的结果。你的任务是首先寻找到重复出现的整数，再找到丢失的整数，将它们以数组的形式返回。
 *
 * 示例 1:
 *
 *
 * 输入: nums = [1,2,2,4]
 * 输出: [2,3]
 *
 *
 * 注意:
 *
 *
 * 给定数组的长度范围是 [2, 10000]。
 * 给定的数组是无序的。
 *
 *
 */

// @lc code=start
var findErrorNums = function (nums: number[]): number[] {
  let [dup, missing] = [-1, -1];
  for (let i = 0; i <= nums.length; i++) {
    let count = 0;
    for (let j = 0; j < nums.length; j++) {
      if (nums[j] === i) {
        count++;
      }
    }
    if (count === 2) {
      dup = i;
    } else if (count === 0) {
      missing = i;
    }

    // optimization
    if (dup > 0 && missing > 0) break;
  }
  return [dup, missing];
};

// sorted
var findErrorNums = function (nums: number[]): number[] {
  nums.sort((a, b) => a - b);
  let [dup, missing] = [-1, 1]; // missing = 1
  for (let i = 1; i < nums.length; i++) {
    if (nums[i] === nums[i - 1]) {
      dup = nums[i];
    } else if (nums[i] > nums[i - 1] + 1) {
      missing = nums[i - 1] + 1;
    }
  }

  return [dup, nums[nums.length - 1] !== nums.length ? nums.length : missing];
};

// map
var findErrorNums = function (nums: number[]): number[] {
  const map = new Map<number, number>();
  let [dup, missing] = [-1, -1];

  for (const num of nums) {
    map.set(num, (map.get(num) || 0) + 1);
  }

  for (let i = 1; i <= nums.length; i++) {
    if (map.has(i)) {
      if (map.get(i) === 2) {
        dup = i;
      }
    } else {
      missing = i;
    }
  }

  return [dup, missing];
};

// array
var findErrorNums = function (nums: number[]): number[] {
  const arr = new Array<number>(nums.length + 1).fill(0);
  let [dup, missing] = [-1, -1];
  for (const num of nums) {
    arr[num] += 1;
  }

  for (let i = 1; i < arr.length; i++) {
    if (arr[i] === 0) {
      missing = i;
    } else if (arr[i] === 2) {
      dup = i;
    }
  }

  return [dup, missing];
};

var findErrorNums = function (nums: number[]): number[] {
  let [dup, missing] = [-1, 1]; // missing = 1
  for (const num of nums) {
    if (nums[Math.abs(num) - 1] < 0) {
      dup = Math.abs(num);
    } else {
      nums[Math.abs(num) - 1] *= -1;
    }
  }

  for (let i = 1; i < nums.length; i++) {
    if (nums[i] > 0) {
      missing = i + 1;
    }
  }

  return [dup, missing];
};

// bit manipulation
var findErrorNums = function (nums: number[]): number[] {
  let [xor, xor0, xor1] = [0, 0, 0];

  for (const num of nums) {
    xor ^= num;
  }

  for (let i = 1; i <= nums.length; i++) {
    xor ^= i;
  }

  let rightMostBit = xor & ~(xor - 1);
  for (const num of nums) {
    if ((num & rightMostBit) !== 0) {
      xor1 ^= num;
    } else {
      xor0 ^= num;
    }
  }

  for (let i = 1; i <= nums.length; i++) {
    if ((i & rightMostBit) !== 0) {
      xor1 ^= i;
    } else {
      xor0 ^= i;
    }
  }

  for (let i = 0; i < nums.length; i++) {
    if (nums[i] === xor0) {
      return [xor0, xor1];
    }
  }

  return [xor1, xor0];
};
// @lc code=end

File: leetcode/sorting/645.错误的集合.go
/*
 * @lc app=leetcode.cn id=645 lang=golang
 *
 * [645] 错误的集合
 *
 * https://leetcode-cn.com/problems/set-mismatch/description/
 *
 * algorithms
 * Easy (43.72%)
 * Likes:    215
 * Dislikes: 0
 * Total Accepted:    60.9K
 * Total Submissions: 139.2K
 * Testcase Example:  '[1,2,2,4]'
 *
 * 集合 s 包含从 1 到 n 的整数。不幸的是，因为数据错误，导致集合里面某一个数字复制了成了集合里面的另外一个数字的值，导致集合 丢失了一个数字 并且
 * 有一个数字重复 。
 *
 * 给定一个数组 nums 代表了集合 S 发生错误后的结果。
 *
 * 请你找出重复出现的整数，再找到丢失的整数，将它们以数组的形式返回。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：nums = [1,2,2,4]
 * 输出：[2,3]
 *
 *
 * 示例 2：
 *
 *
 * 输入：nums = [1,1]
 * 输出：[1,2]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 2
 * 1
 *
 *
 */

package leetcode

import "sort"

// @lc code=start
// hash table
func findErrorNums2(nums []int) []int {
	cnt := map[int]int{}
	for _, v := range nums {
		cnt[v]++
	}
	ret := make([]int, 2)
	for i := 1; i <= len(nums); i++ {
		if cnt[i] != 0 {
			if cnt[i] == 2 {
				ret[0] = i
			}
		} else {
			ret[1] = i
		}
	}
	return ret
}

// sort
func findErrorNums(nums []int) []int {
	n := len(nums)
	ret := make([]int, 2)
	sort.Ints(nums)
	pre := 0
	for _, v := range nums {
		if v == pre {
			ret[0] = pre
		} else if v-pre > 1 {
			ret[1] = pre + 1
		}
		pre = v
	}

	if nums[n-1] != n {
		ret[1] = n
	}

	return ret
}

// @lc code=end
